pkg/crawler/crawler.go

```diff
@@ -462,30 +471,54 @@ func (c *Crawler) prepareClassifierData(ctx context.Context) ([]licenseFilesMeta
 	log.Println("Preparing license classifier data")
 
 	var licenseFiles []licenseFilesMeta
+	limit := semaphore.NewWeighted(c.opt.Limit)
 
 	// switch from concurrent to normal map
 	uniqLicenseKeyMap := c.uniqueLicenseKeys.Items()
 	uniqueLicenseKeyList := c.uniqueLicenseKeys.Keys()
-
-	client := http.Client{
-		Timeout: 10 * time.Second,
+	if len(uniqueLicenseKeyList) == 0 {
+		return licenseFiles, nil
 	}
+	log.Printf("Total license keys to be processed %d", len(uniqueLicenseKeyList))
 
 	licenseKeyChannel := make(chan string, len(uniqueLicenseKeyList))
+	errCh := make(chan error)
+	defer close(errCh)
 
-	log.Printf("Total license keys to be processed %d", len(uniqueLicenseKeyList))
+	// set max timeout for each request processed by the http client for license Meta URLs
+	c.http.HTTPClient.Timeout = 10 * time.Second
+
+	// Set custom request LogHook as needed
+	c.http.RequestLogHook = func(logger retryablehttp.Logger, req *http.Request, attempt int) {
+		if attempt != c.http.RetryMax {
+			return
+		}
+
+		// only on last attempt, we set the request headers to mimick the browser
+		// Reason: Certain URLs ex: https://www.snmp4j.org/GPL.txt, https://www.json.org/license.html
+		// give 503 error response when programatically accessed but when accessed from browser, we get valid response
+		// so to avoid these, we mimick the URLs as if URLs are sent from browser
+
+		req.Header.Set("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/103.0.0.0 Safari/537.36")
+		req.Header.Set("Accept", "text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8")
+		req.Header.Set("Accept-Language", "en-US,en;q=0.5")
+		req.Header.Set("Accept-Encoding", "gzip, deflate, br")
+		req.Header.Set("Connection", "keep-alive")
+		req.Header.Set("Upgrade-Insecure-Requests", "1")
+		req.Header.Set("Cache-Control", "no-cache")
+	}
+
+	defer func() {
+		// reset the values set above
+		c.http.HTTPClient.Timeout = 0
```

Comment:
Aren't we concurrently using the same crawler object? Won't this be a problem?

Reply:
Should not be a problem. the license classification is done at the end of Crawl API. Also regarding the license fetch in different go routines, we are closing the license channel first, next return, next the defer is executed. Do you think this can cause a problem? @namandf

Reply:
But this isn't for classification. This is for preparing the classifier data i.e saving the files locally with license text. That happens as part of the crawl using goroutines. Right?

Reply:
So if multiple goroutines are using the same crawler object and are updating the http client metadata, that might be a problem .

Reply:
No at the end of Crawl, we call the classify license. this then calls the prepareClassifierData. So this happens in main thread only.